### ISSUE TITLE
about Learning rate adjustment

### ISSUE THREAD

**Issue by zcswdt** [ultralytics/yolov5#2055]
## ❔Question
Hello, I drew out the learning rate adjustment parameters of the model and found that there are two learning rates that are completely coincident, but when looking at the code, some add momentum and some weights. Shouldn't they be three different curves? ?

![image](https://user-images.githubusercontent.com/43515926/105976042-a4d4c580-60ca-11eb-997c-7b60d8f1b5c4.png)

**Comment by glenn-jocher** [comment]
@zcswdt thats correct, there are 3 different param groups specified here, each of which may have independent warm and schedulers. These values are also automatically logged in Tensorboard and W&B if you use these loggers.
https://github.com/ultralytics/yolov5/blob/f59f80114cbe4a7b3d4ea743b771aa1919fa8c8a/train.py#L102-L110

**Comment by zcswdt** [comment]
> thats correct, there are 3 different param groups specified here, each of which may have independent warm and schedulers. These values are also automatically logged in Tensorboard and W&B if you use these loggers.

Thank you for your reply. There are some parameters, but two curves completely overlap, so is it a bit redundant to make three sets of parameters?
https://github.com/ultralytics/yolov5/blob/f639e14e4d20f8eaadfbe92dbc766fee7577c56b/train.py#L114-L117

**Comment by glenn-jocher** [comment]
@zcswdt 3 param groups are the minimum required for 3 degrees of freedom, which is what the YOLOv5 training design requires, so no, there is zero redundancy in the this region of the code.

**Comment by zcswdt** [comment]
> 3 param groups are the minimum required for 3 degrees of freedom, which is what the YOLOv5 training design requires, so no, there is zero redundancy in the this region of the code.

I'm sorry I still have some doubts. There are different learning rate adjustment methods for different model layer parameters. Three methods are set here, that is, there are three parameter values. The first one adds momentum, the second adds weight_decay, and the third adds nothing. Shouldn't three different curves be drawn? But the two drawn completely overlap, I still don’t understand in this place, I hope you can give some guidance, thank you!

![image](https://user-images.githubusercontent.com/43515926/106083768-05f6aa80-6158-11eb-8cb0-8701dbeec98e.png)

**Comment by glenn-jocher** [comment]
@zcswdt you're looking at a 1 dimensional chart of param group variables, there are other variables to consider.

**Comment by zcswdt** [comment]
> you're looking at a 1 dimensional chart of param group variables, there are other variables to consider.

Sorry, I still don’t understand. Can you elaborate a little bit? Thanks a lot

**Comment by glenn-jocher** [comment]
See https://pytorch.org/docs/stable/optim.html#torch.optim.SGD

**Comment by zcswdt** [comment]
> See https://pytorch.org/docs/stable/optim.html#torch.optim.SGD

Thank you for your patience and guidance,but I have another question. When you use the warm-up learning rate, the process uses the linear interpolation method. I don’t understand whether the warm-up process is a straight line fitting or whether it can provide some linear interpolation to fit. Additional information or knowledge about the warm-up process is very much appreciated.   
But I have a doubt. Is it possible to use the straight line  y = kx+b  function to fit the warm-up process according to the number of training iterations?

https://github.com/ultralytics/yolov5/blob/f639e14e4d20f8eaadfbe92dbc766fee7577c56b/train.py#L275

**Comment by glenn-jocher** [comment]
@zcswdt warmup operates as intended, it slowly ramps initial LRs to meet their schedulers over a given time period, specified in the hyperparameters.

**Comment by zcswdt** [comment]
> @zcswdt warmup operates as intended, it slowly ramps initial LRs to meet their schedulers over a given time period, specified in the hyperparameters.

Is there any information about linear interpolation methods available?

**Comment by glenn-jocher** [comment]
@zcswdt I'm not sure I follow. Linear interpolation functions are available in most all data science package (numpy, scipy, pytorch etc.), they simply connect two points.

**Comment by zcswdt** [comment]
> @zcswdt I'm not sure I follow. Linear interpolation functions are available in most all data science package (numpy, scipy, pytorch etc.), they simply connect two points.

I understand the linear interpolation formula, but what I don’t understand is the use here, because the warm-up phase I have been in touch with before is the linear adjustment of the number of training iterations and the number of warm-up steps. You use the interpolation method here. , I simulated your method with code, but there are still many doubts.
for example：
![image](https://user-images.githubusercontent.com/43515926/106222472-8c73c080-621a-11eb-9391-468f9c3a113e.png)

**Comment by glenn-jocher** [comment]
@zcswdt the interp functions are used in warmup to smoothly ramp initial LRs to meet their schedulers over a given time period, specified in the hyperparameters. You've plotted this behavior yourself in your original post.
